lib/Sanger/CGP/NgsQc/ValidateSampleMetaAndBamHeader.pm

```diff
@@ -0,0 +1,755 @@
+    if ($bam_state !~ m/^raw$|^pp\-remapped$/) {
+      $traffic_light = 0;
+      last;
+    }
+  }
+
+  ### generate UUIDs and md5sums
+  if ($traffic_light == 1) {
+    #start to produce UUIDs and md5sums
+    @out = ();
+    push @out, $header_line."\t".$HEADER_BAM_STATE."\t".$HEADER_DONOR_UUID."\t".$HEADER_TISSUE_UUID."\t".$HEADER_SAMPLE_UUID."\t".$HEADER_BAM_MD5;
+    my (%exist_donor_uuid, %exist_tissue_uuid);
+    for (my $i = 0;$i < scalar @bam_files; $i++) {
+      my ($donor_uuid, $tissue_uuid, $sample_uuid, $md5sum);
+      if (! exists $exist_donor_uuid{$donor_ids[$i]}) {
+        warn "--------\ngenerating donor UUID for: $donor_ids[$i]\n";
+        $donor_uuid = get_uuid();
+        $exist_donor_uuid{$donor_ids[$i]} = $donor_uuid;
+      } else {
+        warn "--------\nuse existing donor UUID for: $donor_ids[$i]\n";
+        $donor_uuid = $exist_donor_uuid{$donor_ids[$i]};
+      }
+
+      if (! exists $exist_tissue_uuid{$tissue_ids[$i]}) {
+        warn "generating tissue UUID for: $tissue_ids[$i]\n";
+        $tissue_uuid = get_uuid();
+        $exist_tissue_uuid{$tissue_ids[$i]} = $tissue_uuid;
+      } else {
+        warn "use existing tissue UUID for: $tissue_ids[$i]\n";
+        $tissue_uuid = $exist_tissue_uuid{$tissue_ids[$i]};
+      }
+      warn "generating sample UUID for: $sample_ids[$i]\n";
+      $sample_uuid = get_uuid();
+
+      my $left = scalar @bam_files - $i - 1;
+      warn "generating md5sum for: $bam_files[$i]. [$left bam(s) to go]\n";
+      $md5sum = get_md5sum($bam_files[$i]);
+      push @out, $lines[$i]."\t".$bam_states[$i]."\t".$donor_uuid."\t".$tissue_uuid."\t".$sample_uuid."\t".$md5sum;
+      warn "done!\n--------\n";
+    }
+  }
+
+  if ($opts->{'mod'} == 1) {
+    write_results(\@out, $opts->{'o'}, $opts->{'f'}, $traffic_light);
+    return ($traffic_light);
+  } else { # return array ref when in test mode, instead of writing to file
+    return (\@out, $traffic_light);
+  }
+}
+
+sub get_md5sum {
+  my $file = shift;
+  open(FILE, "<", $file);
+  my $cksum = md5_hex(<FILE>);
+  if ($cksum) {
+    return $cksum;
+  } else {
+    return 0;
+  }
+}
+
+sub get_uuid {
+  my $ug = Data::UUID->new;
+  my $uuid = $ug->to_string($ug->create);
+  return $uuid;
+}
+
+sub validate_bam {
+  my ($sample_ID, $is_normals, $align_file, $check_all, $count_base_number) = @_; # $count_base_number is for testing the script
+  my $bam_state = 'failed:';
+
+  if ($is_normals !~ m/^yes$|^y$|^no$|^n$/i) {
+    $bam_state .= $FAIL_CODE_010."; ";
+  }
+
+  if (! -e $align_file) {
+    warn "can not find the file: $align_file\n";
+    return $bam_state.$FAIL_CODE_020.";";
+  }
+
+  warn "--------\nstart to validate alignment file: $align_file\n--------\n";
+  my $sam = Bio::DB::HTS->new(-bam => $align_file);
+  my @lines = split "\n", $sam->header->text;
+  my %headers; # to handle duplicates!
+
+  my (@RG_lines, @SQ_lines, @PG_lines);
+  for(@lines) {
+    push @RG_lines, $_ if(index($_,'@RG') == 0);
```

Comment:
You'll find that if you are anchoring a search to the start that a regex match is faster.  This is trivial in small scripts but worth knowing.  Basically `index` tries to find an instance of the specified string in the whole variable and then returns a value.  A regex will exit early if a condition is fulfilled, in this case within the first 3 characters.

Efficiency wise you are actually better here doing something like:

```perl
my %header_sets;
for(@lines) {
  chomp $_;
  my ($type, @bits) = split /\t/, $_;
  my %fields;
  for my $bit(@bits) {
    my ($ft, $val) = $bit =~ m/([^:]+):(.+)/;
    $fields{$ft}=$val;
  }
  $type =~ s/^\@//;
  push $header_sets{$type}, \%fields;
}
```

Now you can just do exist's tests on the hashes in the next section, if we decided to test an additional header key this section wouldn't need to change.

Hopefully this makes sense